# Hands-on Exercise for  FPM Module

### 1. Exploring properties of the dataset accidents_10k.dat. Read more about it here:  http://fimi.uantwerpen.be/data/accidents.pdf

In [1]:
!head accidents_10k.dat

1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 
2 5 7 8 9 10 12 13 14 15 16 17 18 20 22 23 24 25 27 28 29 32 33 34 35 36 37 38 39 
7 10 12 13 14 15 16 17 18 20 25 28 29 30 33 40 41 42 43 44 45 46 47 48 49 50 51 52 
1 5 8 10 12 14 15 16 17 18 19 20 21 22 24 25 26 27 28 29 30 31 41 43 46 48 49 51 52 53 54 55 56 57 58 59 60 61 
5 8 10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 31 33 36 38 39 41 43 46 56 62 63 64 65 66 67 68 
7 8 10 12 17 18 21 23 24 26 27 28 29 30 33 34 35 36 38 41 43 47 59 63 66 69 70 71 72 73 74 75 76 77 78 79 
1 12 14 15 16 17 18 21 22 23 24 25 27 28 29 30 31 35 38 41 43 44 53 56 57 58 59 60 63 66 80 81 82 83 84 
10 12 14 15 16 17 18 21 22 24 25 26 27 28 29 30 31 33 39 41 43 44 46 49 59 60 62 63 66 82 
1 8 10 12 14 15 16 17 18 21 22 23 24 25 27 29 30 31 38 41 43 53 56 59 61 63 66 68 85 86 87 88 89 
1 8 12 13 14 15 16 17 18 22 24 25 28 30 38 41 42 43 46 49 60 63 64 66 80 82 84 90 91 92 93 94 95 


<span style="color:red">**Question 1a:** </span>. How many items are there in the data?

In [2]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' accidents_10k.dat 

310


In [3]:
##!awk -- '{for (i = 1; i <= length; i++) wc += 1}; END {print wc}' accidents_10k.dat 

<span style="color:green">**Answer:** </span> 310 items are there in the data.

<span style="color:red">**Question 1b:** </span> How many transactions are present in the data?

In [4]:
!wc -l accidents_10k.dat

10000 accidents_10k.dat


<span style="color:green">**Answer:** </span> 10,000 transactions

<span style="color:red">**Question 1c:** </span>.  What is the length of the smallest transaction?

In [5]:
##!awk '{print $NF}' accidents_10k.dat

In [6]:
##!awk "{print length}" accidents_10k.dat

In [7]:
!awk '(NR==1||length<s){s=length} END {print s}' accidents_10k.dat

69


<span style="color:green">**Answer:** </span>69

<span style="color:red">**Question 1d:** </span>  What is the length of the longest transaction?

In [8]:
!awk '{ if (length > max) max = length } END { print max }' accidents_10k.dat 

145


<span style="color:green">**Answer:** </span> 145

<span style="color:red">**Question 1e:** </span>  What is the size of the search space of frequent itemsets in this data?

In [9]:
print(pow(2,310))

2085924839766513752338888384931203236916703635113918720651407820138886450957656787131798913024


<span style="color:green">**Answer:** </span> 2 power 310

<span style="color:red">**Question 1f:** </span> 
Assume that you work for the deparment of transportation that collected this data. What benefit do you see in using itemset mining approaches on this data?

<span style="color:green">**Answer:** </span> By using this approach we can find the possible accidents in the future. I can use the data and predict the possible accident locations and the type of roads most of the accidents occur. I Can also filter the possible accidents based on the season by seeing the common pattern of accidents in the previous seasons/year.

<span style="color:red">**Question 1g:** </span>  What type of itemsets (frequent, maximial or closed) would you be interested in discovering this dataset? State your reason.

<span style="color:green">**Answer:** </span> I would be interested to find the number of Maximal Itemsets in the dataset. I want to find the frequent itemsets but i dont want huge chunk of frequent itemsets, so i will find maximum itemsets(which can be useful to derive all the maximal itemsets).<br><br> But if i want to find the most occuring or best frequent pattern i will find the closed itemsets.

<span style="color:red">**Question 1h:** </span>  What minsup threshold would you use and why?

<span style="color:green">**Answer:** </span> If i take very low minsup then the number of frequent itemsets in the data can be very large. If i take high minsup then i could miss out on some important frequent pattern in the data. So i should take minsup such that it captures best possib le frequent itemsets. I would choose minsup = Total_transactions/2 or Total_transactions/3. That is in this case 4000 or 5000.<br><br> The selection of minsup purely depends on the data and how i want to capture the data.

### 2. Generating frequent, maximal and closed itemsets using $\color{red}{\text{Apriori}}$, $\color{red}{\text{ECLAT}}$, and $\color{red}{\text{FPGrowth}}$ algorihtms from the dataset accidents_10k.dat 

<span style="color:red">**Question 2a:** </span> Generate frequent itemsets using Apriori, for minsup = 2000, 3000, and 4000. Which of these minsup thresholds results in a maximum number of frequent itemsets? Which of these minsup thresholds results in a least number of frequent itemsets? Provide a rationale for these observations.

In [10]:
!chmod u+x apriori

In [11]:
!./apriori

usage: ./apriori [options] infile [outfile]
find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
-t#      target type                              (default: s)
         (s: frequent, c: closed, m: maximal item sets,
          g: generators, r: association rules)
-m#      minimum number of items per set/rule     (default: 1)
-n#      maximum number of items per set/rule     (default: no limit)
-s#      minimum support of an item set/rule      (default: 10%)
-S#      maximum support of an item set/rule      (default: 100%)
         (positive: percentage, negative: absolute number)
-o       use original rule support definition     (body & head)
-c#      minimum confidence of an assoc. rule     (default: 80%)
-e#      additional evaluation measure            (default: none)
-a#      aggregation mode for evaluation measure  (default: none)
-d#      threshold for add. evaluation measure    (default: 10%)
-i       invalidate eva

In [12]:
!./apriori -ts -s-2000 accidents_10k.dat accidents_minsup2000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [16.98s].
writing accidents_minsup2000.txt ... [851034 set(s)] done [0.09s].


In [13]:
!./apriori -ts -s-3000 accidents_10k.dat accidents_minsup3000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.00s].
building transaction tree ... [24741 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 done [4.12s].
writing accidents_minsup3000.txt ... [133799 set(s)] done [0.01s].


In [14]:
!./apriori -ts -s-4000 accidents_10k.dat accidents_minsup4000.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.01s].
building transaction tree ... [22267 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.14s].
writing accidents_minsup4000.txt ... [29501 set(s)] done [0.00s].


In [15]:
!wc -l accidents_minsup2000.txt
!wc -l accidents_minsup3000.txt
!wc -l accidents_minsup4000.txt

851034 accidents_minsup2000.txt
133799 accidents_minsup3000.txt
29501 accidents_minsup4000.txt


<span style="color:green">**Answer:** </span> For minsup=2000: [851034 set(s)] 
For minsup=3000: [133799 set(s)] 
For minsup=4000: [29501 set(s)] .<br> So we get maximum number of frequent itemsets for minsup=2000.<br>
And we get maximum number of frequent itemsets for minsup=4000.<br>
<br> From this we can observe that an increase in minsup can result in getting less number of frequent itemsets. So The minsup and number of frequent itemsets are inversely proportinal to each other.                

<span style="color:red">**Question 2b:** </span>   Using Apriori, compare the execution time for finding frequent itemsets for minsup = 2000, 3000, and 4000. Which of these minsup thresholds takes the least amount of time? Provide a rationale for this observation.

In [16]:
import datetime 
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat 
end = datetime.datetime.now()
elapsed = end - start
print("Time: ",elapsed);
print("------------------------------------")
start2 = datetime.datetime.now()
!./apriori -ts -s-3000 accidents_10k.dat 
end2 = datetime.datetime.now()
elapsed2 = end2 - start2
print("Time: ",elapsed2);
print("------------------------------------")
start3 = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat
end3 = datetime.datetime.now()
elapsed3 = end3 - start3
print("Time: ",elapsed3);
print("------------------------------------")

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.00s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [17.06s].
writing <null> ... [851034 set(s)] done [0.01s].
Time:  0:00:17.709695
------------------------------------
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.03s].
filtering, sorting and recoding items ... [38 item(s)] done [0.00s].
sorting and reducing transactions ... [9674/10000 transaction(s)] done [0.01s].
building transaction tree ... [24741 node(s)

<span style="color:green">**Answer:** </span>For minsup=2000 the execution time is greater than minsup=4000.
    <br>Execution time is in the order: minsup2000 > minsup3000 > minsup4000.
    <br> The least time to run the apriori algorithm is when minsup = 4000.
    <br> If the support value is higher it results in less number of branches in the nodes of the tree. Then it results in giving less number of branches in the next level. It means looking in less number of nodes in the tree, which results in faster execution.

<span style="color:red">**Question 2c:** </span> Using Apriori, find the frequent itemsets for minsup = 2000, 3000, and 4000. Determine the number of itemsets for each size (1 to max length of an itemset). What trends do you see that are common for all three minsup thresholds? What trends do you see that are different? Provide a rationale for these observations.

In [17]:
!awk '{print NF-1}' accidents_minsup2000.txt|sort -n|uniq -c

     49 1
    705 2
   5285 3
  23745 4
  69647 5
 139628 6
 195730 7
 193299 8
 133819 9
  63937 10
  20497 11
   4189 12
    483 13
     21 14


In [18]:
!awk '{print NF-1}' accidents_minsup3000.txt|sort -n|uniq -c

     38 1
    468 2
   2830 3
   9887 4
  21779 5
  31964 6
  32020 7
  21862 8
   9839 9
   2705 10
    387 11
     20 12


In [19]:
!awk '{print NF-1}' accidents_minsup4000.txt|sort -n|uniq -c

     33 1
    319 2
   1492 3
   4043 4
   6926 5
   7751 6
   5626 7
   2546 8
    668 9
     91 10
      6 11


<span style="color:green">**Answer:** </span> Initially The number of Itemsets are low but its number increases and after the meanpoint of the frequent itemsets, decreases till the max length of itemset.<br><br> It is like a normal distribution curve. <br>
The max length of Frequent Itemset decreases with the increase in the minsup.
<br> Even if the number of frequent itemsets decreases the pattern follows a normal distribution,where most of the frequent itemsets are in the median length of itemsets.

<span style="color:red">**Question 2d:** </span>  Using Apriori with minsup=2000, compare the number of frequent, maximal, and closed itemsets. Which is the largest set and which is the smallest set? Provide a rationale for these observations.

In [20]:
!./apriori -tm -s-2000 accidents_10k.dat maximal_itemset.txt
!./apriori -tc -s-2000 accidents_10k.dat closed_itemset.txt

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 14 done [24.87s].
filtering for maximal item sets ... done [0.02s].
writing maximal_itemset.txt ... [12330 set(s)] done [0.01s].
./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].
building transaction tree ... [20250 n

<span style="color:green">**Answer:** </span> Number of Frequent Itemsets: 851034;
<br> Number of Maximal Itemsets: 12330; 
<br> Number of Closed Itemsets: 519902;
<br> Largest Set: Frequent Itemset; Smallest Set: Maximal Itemset.
<br> The maximal itemsets are subset of closed itemsets which are subset of frequent itemsets.
<br> Maximal Itemsets &sub; Closed Itemsets &sub; Frequent Itemset.

<span style="color:red">**Question 2e:** </span> For a minsup = 2000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [21]:
!chmod u+x fpgrowth

In [22]:
!chmod u+x eclat

In [23]:
start = datetime.datetime.now()
!./apriori -ts -s-2000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print("------------------------------------")
print("Time for apriori: ",elapsed);
print("------------------------------------")
start2 = datetime.datetime.now()
!./eclat -ts -s-2000 accidents_10k.dat
end2 = datetime.datetime.now()
elapsed2 = end2 - start2
print("------------------------------------")
print("Time for eclat: ",elapsed2);
print("------------------------------------")
start3 = datetime.datetime.now()
!./fpgrowth -ts -s-2000 accidents_10k.dat
end3 = datetime.datetime.now()
elapsed3 = end3 - start3
print("------------------------------------")
print("Time for fpgrowth: ",elapsed3);

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [49 item(s)] done [0.01s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.00s].
building transaction tree ... [20250 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 12 13 done [17.01s].
writing <null> ... [851034 set(s)] done [0.01s].
------------------------------------
Time for apriori:  0:00:17.670756
------------------------------------
./eclat - find frequent item sets with the eclat algorithm
version 5.21 (2019.03.31)        (c) 2002-2019   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [49 item(s)] done [0.00s].
sorting and reducing transactions ... [9951/10000 transaction(s)] done [0.01s].

<span style="color:green">**Answer:** </span> 
FPgrowth algorithm took least amount of time compared to other two algorithms.
<br> It is a projection baswed approach where the number of nodes are greatly reduced and we check each node whether the item satisfy the minsup condition or not. This can greatly reduce the search space and time of execution.<br> Then the best approach is the ECLAT algorithm which is based on the vertical approach. The apriori algorithm on the other hand is a level wise approach which takes a lot of time with an increase in itemsets.

<span style="color:red">**Question 2f:** </span> For a minsup = 4000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [24]:
start = datetime.datetime.now()
!./apriori -ts -s-4000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print("------------------------------------")
print("Time for apriori: ",elapsed);
print("------------------------------------")
start2 = datetime.datetime.now()
!./eclat -ts -s-4000 accidents_10k.dat
end2 = datetime.datetime.now()
elapsed2 = end2 - start2
print("------------------------------------")
print("Time for eclat: ",elapsed2);
print("------------------------------------")
start3 = datetime.datetime.now()
!./fpgrowth -ts -s-4000 accidents_10k.dat
end3 = datetime.datetime.now()
elapsed3 = end3 - start3
print("------------------------------------")
print("Time for fpgrowth: ",elapsed3);

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
building transaction tree ... [22267 node(s)] done [0.01s].
checking subsets of size 1 2 3 4 5 6 7 8 9 10 11 done [1.21s].
writing <null> ... [29501 set(s)] done [0.00s].
------------------------------------
Time for apriori:  0:00:01.494634
------------------------------------
./eclat - find frequent item sets with the eclat algorithm
version 5.21 (2019.03.31)        (c) 2002-2019   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.02s].
filtering, sorting and recoding items ... [33 item(s)] done [0.00s].
sorting and reducing transactions ... [9381/10000 transaction(s)] done [0.00s].
writing

<span style="color:green">**Answer:** </span>FPgrowth algorithm took least amount of time compared to other two algorithms. The apriori algorithm execution time is greatly reduced compared to when we taken minsup=2000.


<span style="color:red">**Question 2g:** </span>  For a minsup = 6000, compare the execution time for Apriori, ECLAT and FPGrowth. Which of these algorithms took the least amount of time. Provide a rationale for this observation.

In [26]:
start = datetime.datetime.now()
!./apriori -ts -s-6000 accidents_10k.dat
end = datetime.datetime.now()
elapsed = end - start
print("------------------------------------")
print("Time for apriori: ",elapsed);
print("------------------------------------")
start2 = datetime.datetime.now()
!./eclat -ts -s-6000 accidents_10k.dat
end2 = datetime.datetime.now()
elapsed2 = end2 - start2
print("------------------------------------")
print("Time for eclat: ",elapsed2);
print("------------------------------------")
start3 = datetime.datetime.now()
!./fpgrowth -ts -s-6000 accidents_10k.dat
end3 = datetime.datetime.now()
elapsed3 = end3 - start3
print("------------------------------------")
print("Time for fpgrowth: ",elapsed3);

./apriori - find frequent item sets with the apriori algorithm
version 6.27 (2017.08.01)        (c) 1996-2017   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [20 item(s)] done [0.01s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
building transaction tree ... [6478 node(s)] done [0.00s].
checking subsets of size 1 2 3 4 5 6 7 8 done [0.04s].
writing <null> ... [2254 set(s)] done [0.00s].
------------------------------------
Time for apriori:  0:00:00.306524
------------------------------------
./eclat - find frequent item sets with the eclat algorithm
version 5.21 (2019.03.31)        (c) 2002-2019   Christian Borgelt
reading accidents_10k.dat ... [310 item(s), 10000 transaction(s)] done [0.01s].
filtering, sorting and recoding items ... [20 item(s)] done [0.01s].
sorting and reducing transactions ... [3216/10000 transaction(s)] done [0.00s].
writing <null> ..

<span style="color:green">**Answer:** </span> FPgrowth algorithm still takes the least amount of time compared to other two algorithms.<br> But we can see as minsup value increases, the execution times of the given three algorithms are almost closeby/similar.

<span style="color:red">**Question 2h:** </span> Fill the following table based on execution times computed in __2e__, __2f__, and __2g__. State your observations on the relative computational efficiency at different support thresholds. Based on your knowledge of these algorithms, provide the reasons behind your observations.

|   Algorithm                |minsup=2000         |minsup=4000         |minsup=6000         |
|----------------------------|--------------------|--------------------|--------------------|    
|Apriori                     |    17.670756       |       01.494634    |      00.306524     |
|Eclat                       |    00.520795       |       00.332723    |      00.272711     |
|FPGrowth                    |    00.350361       |       00.253348    |      00.273894     |

<span style="color:green">**Answer:** </span>  If we take lower minsup value the cost of execution of the algorithm vary differently. FPgrowth better than ECLAT, which is better than Apriori. <br> But as the value of minsup increases the execution time is decreased greatly in apriori algorithm. This makes the execution time of the the algorithms at high mnsup, almost similar.<br> Also there is not much change in the execution time of the FPGrowth algorithm when we change minsup value.  <br><br> So, if we take minsup high then the cost of execution is simliar in all these algorithms.

### 3. Discovering frequent subsequences and substrings

Assume that roads in a Cincinnati are assigned numbers. Participants are enrolled in a transportation study and for every trip they make using their car, the sequence of roads taken are recorded. Trips that involves freeways are excluded. This data is in the file <span style="color:blue">road_seq_data.dat</span>.

<span style="color:red">**Question 3a:** </span>  What 'type' of sequence mining will you perform to determine frequently taken 'paths'? Paths are sequences of roads traveresed consecutively in the same order.

<span style="color:green">**Answer:** </span> The frequent subsequence mining involves finding items that are common but there is no order in finding these subsequences, so finding the subsequences wont help to find the "paths". But in finding the Substring Mining, there is a specific order we need to find the substrings. There is a window of certain size and we need to find the substring with the same path that is we find the substring with the same "path".<br><br> Therefore, substring sequence mining will perform frequently taken "paths".

<span style="color:red">**Question 3b:** </span> How many sequences are there in this sequence database?

In [27]:
!wc -l road_seq_data.dat

1000 road_seq_data.dat


<span style="color:green">**Answer:** </span> 1000

<span style="color:red">**Question 3c:** </span> What is the size of the alphabet in this sequence database?

In [28]:
!awk -- '{for (i = 1; i <= NF; i++) wc[$i] += 1}; END {print length(wc)}' road_seq_data.dat 

1283


<span style="color:green">**Answer:** </span> 1283

<span style="color:red">**Question 3d:** </span> What are the total number of possible subsequences of length 2 in this dataset?

In [29]:
!chmod u+x prefixspan

In [30]:
!./prefixspan

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei


Usage: prefixspan [OPTION]... INFILE

       where [OPTION]...  is a list of zero or more optional arguments
             INFILE(s)    is the name of the input transaction database

Additional arguments (at most one input file may be specified):
       -min_sup [minimum support]
       -max_pat [maximum pattern]



In [35]:
!./prefixspan -min_sup 1 road_seq_data.dat > road_minsup_1.dat

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei



In [36]:
!head road_minsup_1.dat


( 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255 256 257 258 259 260 261 262 263 264 265 266 267 268 269 270 271 272 273 274 275 276

In [37]:
!awk '{print NF}' road_minsup_1.dat|sort -n|uniq -c

      1 0
   1283 1
  15927 2
  71547 3
 178215 4
1603973 5
 502491 6
 330358 7
 224946 8
 109485 9
  43350 10
  13284 11
   7438 12
   1508 13
   1268 14
    499 15
    406 16
    328 17
    385 18
    218 19
    142 20
    146 21
     75 22
     61 23
     94 24
    165 25
     18 26
     24 27
     42 28
     21 29
     34 30
     76 31
     38 32
     41 33
     79 34
     16 35
     53 36
     20 37
      8 38
     28 39
     40 40
      4 41
     33 42
      2 43
      2 44
      6 46
      7 47
      7 48
      4 49
     19 50
     36 51
     10 53
      8 54
      4 55
      3 56
      3 57
      5 58
      7 59
      4 60
      7 61
      2 62
      3 63
      5 64
      4 65
      3 66
      1 69
      6 72
      4 75
      4 76
      6 78
      1 81
      1 82
      3 83
      5 85
      4 89
      4 92
      6 93
      1 94
      2 95
      2 98
      3 100
      1 101
      2 102
      6 104
      2 108
      1 112
      2 117
      1 118
      2 120
      2 138
      3 14

In [48]:
Answer :: 15927

<span style="color:red">**Question 3e:** </span> What are the total number of possible substrings of length 2 in this dataset?

In [39]:
!./seqwog -ts -s-1 road_seq_data.dat road_min_1.txt

./seqwog - find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.00s].
recoding items ... [1283 item(s)] done [0.00s].
reducing and triming transactions ... [910/1000 transaction(s)] done [0.00s].
writing road_min_1.txt ... [28556 sequence(s)] done [0.03s].


In [40]:
!awk '{print NF-1}' road_min_1.txt|sort -n|uniq -c

   1283 1
   2295 2
   3247 3
   3874 4
   3990 5
   3758 6
   3311 7
   2669 8
   1946 9
   1204 10
    614 11
    267 12
     83 13
     15 14


In [42]:
Answer :: 2295

<span style="color:red">**Question 3f:** </span> Discover frequent __subsequences__ with minsup = 10 and report the number of subsequences discovered.

In [43]:
!chmod u+x sequoia

In [44]:
!./sequoia -ts -s-10 road_seq_data.dat subsequences_10.txt

./sequoia - sequence mining with unique occurrences of items and weight averaging
version 2.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.00s].
recoding items ... [1283 item(s)] done [0.00s].
filtering and reducing transactions ... [910/1000 transaction(s)] done [0.00s].
writing subsequences_10.txt ... [4588 sequence(s)] done [0.03s].


In [45]:
!./prefixspan -min_sup 10 road_seq_data.dat > prefix_minsup10.dat

PrefixSpan version 1.00 - Sequential Pattern Miner
Written by Yasuo Tabei



In [46]:
!wc -l prefix_minsup10.dat

9178 prefix_minsup10.dat


<span style="color:green">**Answer:** </span> 9178

<span style="color:red">**Question 3g:** </span>  Discover frequent __substrings__ with minsup = 10 and report the number of substrings discovered.

In [47]:
!./seqwog -ts -s-10 road_seq_data.dat substrings_10.txt

./seqwog - find frequent sequences without gaps
version 3.16 (2016.10.15)        (c) 2010-2016   Christian Borgelt
reading road_seq_data.dat ... [1283 item(s), 1000 transaction(s)] done [0.00s].
recoding items ... [1283 item(s)] done [0.00s].
reducing and triming transactions ... [844/1000 transaction(s)] done [0.00s].
writing substrings_10.txt ... [613 sequence(s)] done [0.00s].


<span style="color:green">**Answer:** </span> 613 substrings.


<span style="color:red">**Question 3h:** </span> Explain the difference in the number of frequent subsequences and substrings found in __3f__ and __3g__ above.

<span style="color:green">**Answer:** </span> The number of subsequences found for the given dataset is more than the number of substrings found for the same given minsup.